### PR TITLE
KAFKA-3177: log warning when topic/partition doesn't exist

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by omkreddy** [comment]
LGTM

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -461,10 +461,15 @@ private int append(Map<TopicPartition, List<ConsumerRecord<K, V>>> drained,
...
         Map<TopicPartition, ListOffsetRequest.PartitionData> partitions = new HashMap<>(1);
         partitions.put(topicPartition, new ListOffsetRequest.PartitionData(timestamp, 1));
-        PartitionInfo info = metadata.fetch().partition(topicPartition);
+        Cluster cluster = metadata.fetch();
+        PartitionInfo info = cluster.partition(topicPartition);
         if (info == null) {
+            if (!cluster.topics().contains(topicPartition.topic()))
+                log.warn("No metadata available for topic {}, wait for metadata refresh", topicPartition.topic());
+            else
+                log.warn("No metadata available for partition {}, wait for metadata refresh", topicPartition);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Is it worth saying something about `fetching offset` as in the previous message?

**Comment by ijuma** [comment]
Since this is not the full fix for KAFKA-3177, let's either file a new issue or make sure it remains open after we merge this.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 2]
```diff
@@ -461,10 +461,15 @@ private int append(Map<TopicPartition, List<ConsumerRecord<K, V>>> drained,
     private RequestFuture<Long> sendListOffsetRequest(final TopicPartition topicPartition, long timestamp) {
         Map<TopicPartition, ListOffsetRequest.PartitionData> partitions = new HashMap<>(1);
         partitions.put(topicPartition, new ListOffsetRequest.PartitionData(timestamp, 1));
-        PartitionInfo info = metadata.fetch().partition(topicPartition);
+        Cluster cluster = metadata.fetch();
+        PartitionInfo info = cluster.partition(topicPartition);
         if (info == null) {
+            if (!cluster.topics().contains(topicPartition.topic()))
+                log.warn("No metadata available for topic {}, wait for metadata refresh", topicPartition.topic());
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Should this be WARN or INFO? I feel it is not "abnormal" that the topic partition does not exist yet on bootstrap or topic update?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Good point. We could even leave this at DEBUG level if it's not as exceptional as a missing partition.

**Comment by ijuma** [comment]
I thought we had merged this, but looks like we dropped it. @hachikuji, is it worth reviving it?

**Comment by hachikuji** [comment]
@ijuma We may as well get this in. Even if the fix may not satisfy all users, at least problems will be more apparent in the logs.

**Comment by hachikuji** [comment]
Discussed offline with @ijuma. We think this patch may not be providing much additional value at the moment and there's some risk of excess log spam. It's probably better to leave this issue open and come back to it after the dust has settled as far as the long-term handling of unknown topics/partitions. This problem should primarily affect users using manual assignment and they can currently enable DEBUG logging to get more details.

**Comment by ewencp** [comment]
@hachikuji If this patch isn't going to be useful, maybe close the PR? And do you have thoughts on the JIRA? If we're not addressing it in 0.10.2 then I'm going to remove the version. I also suspect we can reduce the priority given that we haven't gotten to it for so long (it started with fix version 0.10.0.0). @ijuma do you have thoughts since you were the one that originally bumped priority from major to critical? I haven't noticed many mailing list questions about this.

**Comment by ijuma** [comment]
@ewencp, I'm OK with reducing the priority and removing the "fix version". The original thinking for increasing the priority is the change we made to return `UnknownTopicOrPartition` in case the user is not authorized to describe the topics. Given feedback so far though, it seems that it's not as bad as I thought it would be.

**Pull request closed without merging**